cranelift/codegen/src/isa/riscv64/inst.isle

```diff
@@ -2262,13 +2285,183 @@
+;; as unsigned. This also holds true when viewing both of these numbers as
+;; signed (-2 < -1), so we can use the unsigned comparison for the bottom half.
+;;
+;; Emit the following sequence:
+;;     slt{,u}   t1, x_hi, y_hi
+;;     sltu      t2, x_lo, y_lo
+;;     beq       x_hi, y_hi, .top_is_equal
+;;     mov       rd, t1
+;;     j         .end
+;; .top_is_equal:
+;;     mov       rd, t2
+;; .end:
+(rule 2 (gen_icmp_inner cc x y $I128)
+  (if-let (IntCC.UnsignedLessThan) (intcc_unsigned cc))
+  (let ((x_lo Reg (value_regs_get x 0))
+        (x_hi Reg (value_regs_get x 1))
+        (y_lo Reg (value_regs_get y 0))
+        (y_hi Reg (value_regs_get y 1))
+        ;; Generate compares for both halves.
+        ;; The bottom compare depends on the IntCC.
+        (top_cmp Reg (gen_icmp_inner cc x_hi y_hi $I64))
+        (bottom_cmp Reg (rv_slt x_lo y_lo)))
+    ;; If the high parts are equal, the result only depends on the bottom
+    (gen_select_reg (IntCC.Equal) x_hi y_hi bottom_cmp top_cmp)))
+
+
+;; Compare both registers using xor, and set the result using the dedicated
+;; `seqz`/`snez` instructions.
+(rule (gen_icmp_inner (IntCC.Equal) x y (fits_in_64 ty))
+  (rv_seqz (rv_xor x y)))
+(rule (gen_icmp_inner (IntCC.NotEqual) x y (fits_in_64 ty))
+  (rv_snez (rv_xor x y)))
+
+;; In the I128 case we just `xor` everything and check if its zero at the end.
+(rule 1 (gen_icmp_inner (IntCC.Equal) x y $I128)
+  (let ((top_eq Reg (rv_xor (value_regs_get x 0) (value_regs_get y 0)))
+        (bottom_eq Reg (rv_xor (value_regs_get x 1) (value_regs_get y 1)))
+        (res Reg (rv_or top_eq bottom_eq)))
+    (rv_seqz res)))
+(rule 1 (gen_icmp_inner (IntCC.NotEqual) x y $I128)
+  (let ((top_eq Reg (rv_xor (value_regs_get x 0) (value_regs_get y 0)))
+        (bottom_eq Reg (rv_xor (value_regs_get x 1) (value_regs_get y 1)))
+        (res Reg (rv_or top_eq bottom_eq)))
+    (rv_snez res)))
+
+
+
+
+;; For some icmp's we can optimize the instruction sequence if the RHS is a constant.
+;;
+;; TODO: Currently we only have rules for <=64bit types. We can add more rules for I128's
+(decl gen_icmp_imm (IntCC ValueRegs i64 Type) Reg)
+
+;; This rule isn't totally necessary since we can just use `slti 0`, but
+;; it's the official mnemonic for this operation and gives a slightly nicer
+;; disassembly output.
+(rule 5 (gen_icmp_imm (IntCC.SignedLessThan) x 0 (fits_in_64 ty))
+  (rv_sltz (sext x ty $I64)))
+
+;; `sgtz` is preferable since our equivalent immediate lowering has to do `slt+xori`.
+(rule 5 (gen_icmp_imm (IntCC.SignedGreaterThan) x 0 (fits_in_64 ty))
+  (rv_sgtz (sext x ty $I64)))
+
+;; For these IntCC's we need to both add 1 to the immediate and invert the result.
+;; i.e. `x > imm` is the same as `!(x < imm + 1)`.
+(rule 4 (gen_icmp_imm cc x imm ty)
+  (if-let (IntCC.UnsignedGreaterThan) (intcc_unsigned cc))
+  (let ((res Reg (gen_icmp_imm (intcc_reverse cc) x (i64_add imm 1) ty)))
+    (rv_xori res (imm12_const 1))))
```

Comment:
Yeah, this case used to be covered when we used Imm12 instead of i64. Since `imm12_add` is partial we would have that automatically checked for us. I forgot to check for that when changing it to i64.